### DISCUSSION TITLE
Simplify constructor method name

### DISCUSSION THREAD

**Discussion by Crown0815** [dotnet/csharplang#5101] [Language Ideas]
Currently, if I have a class and want to add a constructor I have to repeat the name of the class as the method name. This means if I have a class with a really long name like 

```csharp
public class ThisClassCoincidentallyHasAReallyLongName {}
```

adding a constructor requires me to repeat that name. So if I want to add a constructor I have to write

```csharp
public class ThisClassCoincidentallyHasAReallyLongName 
{
    public ThisClassCoincidentallyHasAReallyLongName()
    {
        // Some construction ...
    }
}
```

The issue I see with this is that repeating the class name does not add any additional information to this method (the constructor).

I think that this is a great place for the `this` keyword which is used to reference instance variables and already serves to invoke other constructors. The example above would then turn into 

```csharp
public class ThisClassCoincidentallyHasAReallyLongName 
{
    public this()
    {
        // Some construction ...
    }
}
```

This usage of the `this` keyword would be consistent with the current usage of the `this` keyword in the context of constructors calling each other. Thereby, the class

```csharp
public class SomeClass 
{
    public SomeClass() : this(new parameter)
    {    }

    public SomeClass(object parameter)
    {
        // do something with parameter
    }
}
```

could be written as

```csharp
public class SomeClass 
{
    public this() : this(new parameter)
    {    }

    public this(object parameter)
    {
        // do something with parameter
    }
}
```

**Comment by svick** [comment]
I think this usage of `this` would be very confusing because of indexers:

```c#
public class SomeClass
{
   public this(int i) => // some code
   public this[int i] => // more code
}
```

These two lines look very similar, but do completely different things. If anything was to be done here, I think using the `new` keyword would be much clearer.

**Reply by JustNrik** [reply]
I think `new` may generate confusion as well as `new` is used for shadowing. Though it would be much less confusing than using `this`

**Reply by Crown0815** [reply]
I disagree with `this` being easy to confuse with indexers. First of all, the syntax presented in the example is, to my knowledge, not even valid C#. To use an indexer you need to specify a return type that would turn the example into 

```csharp
public class SomeClass
{
   public this(int i) => // some code
   public object this[int i] => // more code
}
```

Secondly, the style used is rather bad, since while it makes sense to name a variable `i` in the context of an indexer, it is a very bad choice for a constructor parameter.

Thirdly, considering the following class (already valid C#)

```csharp
public class SomeClass
{
   public SomeClass(int someInteger) => // some code
   public object this[int i] => // more code
}
```

Even assuming that the example syntax would be correct, it is hardly more confusing than the already valid

```csharp
var x = new SomeClass(5);  // creates a new instance with someInteger = 5
var y = new SomeClass[5]; // creates a new array with space for 5 entries
```

which creates an instance and an array.

I agree that in the rare case of a poorly named constructor parameter the `this` keyword could be mixed up with an indexer, but it is highly unlikely.

The main argument I see for the `this` keyword over the `new` keyword is that it is already used to refer to constructors in the case of changed constructor calls. Calling the constructors `this` also in their declaration would therefore be consistent (see my example in the original post)

**Reply by svick** [reply]
You're right, I didn't get the syntax right. But my point wasn't that people would literally think this syntax is for indexers, but that the language shouldn't use the same keyword in similar contexts for completely different purposes.

Also, having something that's already confusing in the language is not a justification for adding new confusing things.

**Reply by Crown0815** [reply]
@svick But the keyword `this` is already used to refer to constructors, just not in their declaration but when referencing them from other constructors. Is it not weirder to explain to someone why this syntax 

```
public class SomeClass 
{
    public SomeClass() : this(new object())  {}
    public SomeClass(object parameter) => // construct
}
```
is correct? Would you not expect something like

```
public SomeClass() : SomeClass(new object()) {}
```
or
```
public SomeClass() : new(new object()) {}
```

We already got the `: this(...)` syntax so why not go all the way and say, "yes, that is how a constructor is declared and referred to":

```
public this() : this(new object()) {}
```

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/discussions/1928

**Comment by CyrusNajmabadi** [comment]
My 2c: there's just not a large enough problem here for this to need solving. 

If this were v1, I'd definitely prefer a different syntax. However, we're 20 years in and the syntax we have is good enough and it's not worth the issues that come from having multiple syntaxes to make a change here.

**Reply by Crown0815** [reply]
I agree that this is not a large problem. I would consider this to be somewhat in the ballpark of [target typed new expressions](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/target-typed-new) which according to documentation have as [main motivations](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/target-typed-new#motivation):

> Allow field initialization without duplicating the type.
> Allow omitting the type when it can be inferred from usage.
> Instantiate an object without spelling out the type.

(where, in my opinion, the first and the last points are actually the same, at least in [the examples given](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/target-typed-new#motivation))

I remember to have heard in one of the arguments for this feature that it is just annoying to have to repeat long type names, therefore, it is nice if that could be avoided. It would say the same argument applies here.

**Reply by CyrusNajmabadi** [reply]
> I agree that this is not a large problem. I would consider this to be somewhat in the ballpark of target typed new expressions 

I disagree. Target typed new is, at minimum, 1-2 orders of magnitude more frequent than constructors are.

**Reply by CyrusNajmabadi** [reply]
> It would say the same argument applies here.

There is no doubt this removes some duplication. However, that's just not sufficient for me. A type has. Tiny handful of constructors. However, types are instantiated an enormous number of times. When looking at improvements like this, we focus on the ones that will have a really large impact to make it worth an the costs that new language features bring.

**Reply by Crown0815** [reply]
I do not know the data on this so you may very well be right about that. Without knowing much about language design or compiler work, I would guess that the implementation of constructors using the already reserved `this` keyword and the unambiguity of this feature, I assume that it would also be about 1-2 orders of magnitude easier to implement, would it not?

**Reply by CyrusNajmabadi** [reply]
> I do not know the data on this so you may very well be right about that. 

Signatures of symbols are generally several of orders of magnitude less written than expressions, statements or references of them.

**Reply by Crown0815** [reply]
I absolutely agree with [this](https://github.com/dotnet/csharplang/discussions/5101#discussioncomment-1220616). And duplicating a classes name for the constructor is definitely not a large issue. Nonetheless, I think it would make the constructors easier to use especially in the context of having multiple constructors.

On the other hand, I see there is a lot of work done to specifically avoid defining constructors, as the short definition syntax for records.
Again, I have no idea about the complexity of implementing something like this (and the added documentation etc.). I see it as one of many ways to add more consistency and simplicity to C#.

**Reply by CyrusNajmabadi** [reply]
> I assume that it would also be about 1-2 orders of magnitude easier to implement, would it not?

Yes. 

However, None of my decision making relates to cost of the implementation. My decision making is about the cost to the ecosystem. Now there are two ways to write constructors, and every part of the ecosystem needs to expand energy to adapt to that. 

A feature needs to be very valuable to justify three costs that come with that. See http://stronglyemergent.com/blog/2013/negative-100-points/ as a quick rundown to help understand this piece better. :-)

**Reply by Crown0815** [reply]
I see your point. It is basically the low-hanging fruit problem, where if you pick all the long-hanging fruit you have no time to go for the actual juicy ones (that help your user/customer the most).

Could you point me at a place where the process of digging a feature up from its `-100` points is well shown? I would like to see the arguments that lead to a feature becoming worth implementing.

**Reply by CyrusNajmabadi** [reply]
If you read the design notes we publish you'll see the results of the debate and design discussions we have where we often with decide something is worth it or not. :-)
